support creation of helm chart repo using the existing PHCR form

debsmita1 opened the pull request:
**JIRA:**
https://issues.redhat.com/browse/ODC-6716

**Solution description:**
- modified the existing helm chart repository create form to now support the addition of cluster scoped chart repositories
- Default scope type is `Namespaced`
- The scope type radio button is hidden for users who do not have access to create either namespaced/cluster scoped Helm Chart Repository

**Test coverage report:**
Unit tests
![ut](https://user-images.githubusercontent.com/22490998/178513632-41887ca5-fc8d-46da-a187-81c72a0a1e77.png)

integration tests
![e2e](https://user-images.githubusercontent.com/22490998/178513640-5bf8f18d-61f6-4bde-95a6-ccdec7371942.png)


**GIF:**
![Peek 2022-07-12 20-17](https://user-images.githubusercontent.com/22490998/178518981-ac6c9cb1-31ac-4e70-a0eb-5e30b570fba8.gif)

/kind feature

debsmita1 commented:
/assign @invincibleJai 
/assign @jerolimov

debsmita1 commented:
/cc @serenamarie125

invincibleJai commented:
@debsmita1 Thanks, looks good.  will test it once

should we add a separator here as we have for role binding

<img width="900" alt="image" src="https://user-images.githubusercontent.com/5129024/178936826-d2a26439-f2e2-4383-871b-10a99e908b62.png">


@serenamarie125 @beaumorley ^^

vikram-raj commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepositoryPage.tsx:
```diff
@@ -0,0 +1,58 @@
+  return (
+    <NamespacedPage variant={NamespacedPageVariants.light} hideApplications disabled>
+      <Helmet data-test="helm-plugin~title Create Helm Chart Repository">
+        <title>Create Helm Chart Repository</title>
```
This needs to be translated.

jerolimov commented:
Also take a look, primary from user perspective :smile: 

1. It adds the name field described in this bug https://bugzilla.redhat.com/show_bug.cgi?id=2106366, a separate PR might make this easier, but its also fine to discuss here.

For me, the field labels doesn't match other forms where "Name" is used for the resource name. 

![image](https://user-images.githubusercontent.com/139310/179066905-fb900934-7d4b-4778-aa4f-3fbc91d4d03c.png)

I would suggest field labels like this. PM/UX might get involved here:

* Name
* (Chart Repository) Display name
* (Chart Repository) Description

2. Disable usage of the repo **in the namespace**

"In namespace" shouldn't be shown for cluster-wide HCR. Maybe we can explain that this is related to the catalog?

Something LIKE: Disable this chart repository. Helm Charts will not be shown in the developer catalog.

3. When pinn the HCRs and open the create button from there the URL doesn't contain a namespace.

But my disabled project dropdown shows a namespace.

When switching to PHCRs and then to YAML the namespace is default instead of the project in my dropdown.

https://user-images.githubusercontent.com/139310/179068098-a3e08e5f-1064-4794-9f33-63f029e69e19.mp4

For HCRs we should read the "active namespace" instead of taking the namespace from the URL.

debsmita1 commented:
> Also take a look, primary from user perspective smile
> 
> 1. It adds the name field described in this bug https://bugzilla.redhat.com/show_bug.cgi?id=2106366, a separate PR might make this easier, but its also fine to discuss here.
> 
> For me, the field labels doesn't match other forms where "Name" is used for the resource name.
> 
> ![image](https://user-images.githubusercontent.com/139310/179066905-fb900934-7d4b-4778-aa4f-3fbc91d4d03c.png)
> 
> I would suggest field labels like this. PM/UX might get involved here:
> 
> * Name
> * (Chart Repository) Display name
> * (Chart Repository) Description
> 
> 2. Disable usage of the repo **in the namespace**
> 
> "In namespace" shouldn't be shown for cluster-wide HCR. Maybe we can explain that this is related to the catalog?
> 
> Something LIKE: Disable this chart repository. Helm Charts will not be shown in the developer catalog.
> 
Updated it to
![Screenshot from 2022-07-15 15-12-43](https://user-images.githubusercontent.com/22490998/179198342-9e74b5d7-57ce-4f20-a673-c65db4149929.png)


> 3. When pinn the HCRs and open the create button from there the URL doesn't contain a namespace.
> 
> But my disabled project dropdown shows a namespace.
> 
> When switching to PHCRs and then to YAML the namespace is default instead of the project in my dropdown.
> 
>  Peek.2022-07-14.21-31.mp4 
> For HCRs we should read the "active namespace" instead of taking the namespace from the URL.

Fixed. PTAL

vikram-raj reviewed (approved):
verified it and works as expected.

serenamarie125 commented:
@debsmita1 can you also show the flow for the following users:

1. user that is not cluster admin but can create ProjectHelmChartRepos 
2. user that is completely non-priv ( our typical developer )

debsmita1 commented:
> @debsmita1 can you also show the flow for the following users:
> 
> 1. user that is not cluster admin but can create ProjectHelmChartRepos
>    I'm assuming that the form doesn't have the radio buttons, but the relevant description for Project scoped will still somehow be displayed?

@serenamarie125 
https://user-images.githubusercontent.com/22490998/179807169-0b6f40d4-bea3-4c8b-85a8-04e8e7a456a4.mov

> 2. user that is completely non-priv ( our typical developer )
>    I'm assuming the tile/action is not available for them
Yes, it will be hidden

debsmita1 commented:
/retest

serenamarie125 reviewed (approved):
Looks great!  Nice job with this @debsmita1

invincibleJai commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepository.tsx:
```diff
@@ -0,0 +1,114 @@
+import { convertToForm, convertToHelmChartRepository } from './helmchartrepository-create-utils';
+import { validationSchema } from './helmchartrepository-validation-utils';
+
+export interface CreateHelmChartRepositoryProps {
```
nit: do we need this to be exported

invincibleJai commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepository.tsx:
```diff
@@ -0,0 +1,114 @@
+    formData: convertToForm(resource),
+  });
+
+  const handleSubmit = (values, actions) => {
```
nit: can we have types for this?

invincibleJai commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepository.tsx:
```diff
@@ -0,0 +1,114 @@
+    if (values.editorType === EditorType.YAML) {
+      try {
+        HelmChartRepositoryRes = safeLoad(values.yamlData);
+        if (!HelmChartRepositoryRes?.metadata?.namespace) {
+          HelmChartRepositoryRes.metadata.namespace = namespace;
+        }
```
this will cause NPE if `HelmChartRepositoryRes` will not be falsy

invincibleJai commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/helmchartrepository-create-utils.ts:
```diff
@@ -0,0 +1,91 @@
+export const convertToForm = (resource: HelmChartRepositoryType) => {
+  return {
+    scope: resource.kind,
+    repoName: resource.metadata?.name ?? '',
+    repoDisplayName: resource.spec.name ?? '',
```
metadata has check `?` but not spec?

invincibleJai reviewed (commented):
Thanks @debsmita1 looks good, some nit's see if they make sense.

invincibleJai commented:
/approve

vikram-raj reviewed (commented):
/lgtm
